### DISCUSSION TITLE
"Virtual Distance"

### DISCUSSION THREAD

**Discussion by QJKX** [jamulussoftware/jamulus#2195] [Ideas]
Where we see ping, have a toggle for  "virtual distance".
So instead of 10ms it would show 3.43 meters (ie 10ms times the speed of sound, 343 m/s)
This should probably have the system latency added in as it's trying to convey a feel to the user, not raw technical information (which is better as a ping - because it means something in a networking context)..

The justification is that I think people have a better intuition for what it feels like to play with someone a certain physical distance away, than a certain time away, when the time is fairly low (1/2 a second is intuitive. 35ms is not.)
 It's also a fun comparison that shows just how magic the software is.

**Comment by rdica** [comment]
The ping time displayed is to the server, not each individual player, so your justification makes no sense.

A server in NYC, a player in Toronto won't know what it feels like playing with another player from Washington DC thats also connected to that server as both are subject to their latency to the server, not each other. Network paths for each player to the server don't have any real world "distance" implications, especially in the cluster*** that is North American networks and routing.

Users can put their locations in their profile information if they so choose. Server operators (of which I am one) can also put location information in their server profiles that get sent to directory servers.

Are you confusing the client/server architecture of jamulus with a p2p architecture like sonobus, where your suggestion might make more sense?

Then there's the KISS principle as I understand the Jamulus devs like to follow...

**Comment by QJKX** [comment]
>Are you confusing the client/server architecture of jamulus with a p2p architecture like sonobus, where your suggestion might make more sense?

I'm confusing it with quake which makes the ping time of every player available both in game, and in the server browser. You're right that Jamulus doesn't do that, which severely reduces the value of the idea (although knowing the upper limit of your experience still has value.)

>A server in NYC, a player in Toronto won't know what it feels like playing with another player from Washington DC that's also connected to that server as both are subject to their latency to the server, not each other. 

If you know the ping time of both players (and the various buffers/processing delays) you can work it out as a distance sound would travel in the same time - which is what it feels like.

>especially in the cluster*** that is North American networks and routing.

:-(

**Reply by rdica** [reply]
I do envy the Europeans for this, due to their geography and network policy/design, there are more people, closer together in latency, more servers to choose from, and more participants playing than what we experience here in North America. I listen on some of the EU servers during the day, and love what I hear, wish I could participate, ah well...

**Reply by dcorson-ticino-com** [reply]
And note that because Jamulus synchronises the streams at the server it is possible to play together (after getting used to it) at ping times (distances) that no-one can really imagine.  I play regularly at distances according to your calculation of over 28 meters.
(note also that we have tried using Jamkazam which gives with it's p2p architecture about half the ping time, but as there is no synchronisation it is impossible to really play together even at half the ping time.  For p2p to work one needs really low ping times.)

**Reply by gene96817** [reply]
I continue to say that there are three tiers of latency experiences.
1- many of us have envy for the ultra-low latency apparently experienced by Europeans.
2- In North America (where I used to live), many musicians will drive 2+ hours to a rehearsal. In this environment, mastering latency to rehearse with more distant musicians is another reality. 
3- At the edge of the universe, e.g. in the middle of the Pacific, the low latencies experienced in Europe are impossible. Mastering music with higher latency is a necessity. Sharing music with distant musicians is the holy grail. 

In all of these cases, it is also possible to be connected through one or more overloaded routers that add 20 - 100ms or more. That really distorts the relationship of distance and latency. Please don't only focus on #1. #3 is also wonderful.

**Reply by pcar75** [reply]
I would like to caution all of you to the use of the word 'synchronisation' .  Unless I've missed recent development ,  a Jamulus server does NOT SYNCHRONISES streams ; it MERGES them , without regards latency compensation, timing , tempo or other !!!
I've occasionally had to correct newbies' understanding of Jamulus , possibly confused by the fact that they also use direct monitoring .

**Reply by dcorson-ticino-com** [reply]
You are correct, Jamulus does not synchronise the streams, but through the merging it creates a fixed dephasing that all members hear the same.  Thus synchronisation is possible through player actions.  This possibility of synchronisation is not offered by the p2p architecture.
 So if we don't use the word synchronisation, what word do we use?

**Reply by rdica** [reply]
Merge/mix?

**Reply by gene96817** [reply]
A consistent use of "mix" seems right to me.
Secondarily, we need to continue to emphasize the server does not provide synchronization.

**Reply by dcorson-ticino-com** [reply]
I would rather emphasize that the Jamulus architecture allows the players to synchronise, which the p2p architecture does not.

(I think this is important because I have been playing with groups across the Atlantic on both Jamkazam (p2p) and Jamulus recently.  Jamkazam has half the ping time, but playing together is only possible for very slow and simple pieces as synchronisation is not possible.  I have been very surprised myself at how well Jamulus has allowed us to play even with delays nearing 100ms.  Of course it is the skill of the players making this work, but using Jamkazam the same players have not been able to get such good results even though the ping times are half as much.  I think JK works fine if the ping times are small, but for playing across the Atlantic, for example, things start getting wierd, I have the feeling I am wading through mud, not really playing music.  Everything swims.  Gene do you have comparative experience ? I think you have one of the worst delay time situations I know.)

**Reply by gene96817** [reply]
I have played with Jamulus and JamKazam with musicians between New York and Boston. It is possible to play in sync if you listen to the remote musicians. I abandoned JamKazam for Jamulus because (1) there is more user support in P2P and (2) each P2P node affects performance far more than in a Jamulus (client-server) configuration. 

Yes, I agree with @dcorson-ticino-com's experience. The most impactful observation for me is that the speed of sound in air is roughly 1 ms per foot. Performing on a (large) stage with 40ms is similar to a Jamulus experience. Stop thinking that the latency with Jamulus is so unique and different from in-person music! I take delight in seeing a musician from North America or Europe on an Asian server or someone in Asia on a North American server. Jamulus is awesome. My delay is only half as bad as the experience of a Taiwan musician visiting a North American server.   LOL
